node/src/exchange/server.rs

```diff
@@ -1,52 +1,393 @@
+    tx: mpsc::Sender<(R::Channel, ResponseType)>,
+}
+
+pub(super) trait ResponseSender {
+    type Channel: Send + 'static;
+
+    fn send_response(&mut self, channel: Self::Channel, response: ResponseType);
 }
 
-impl<S> ExchangeServerHandler<S>
+impl ResponseSender for ReqRespBehaviour {
+    type Channel = ResponseChannel<ResponseType>;
+
+    fn send_response(&mut self, channel: Self::Channel, response: ResponseType) {
+        // response was prepared specifically for the request, we can drop it
+        // in case of error we'll get Event::InboundFailure
+        let _ = self.send_response(channel, response);
+    }
+}
+
+impl<S, R> ExchangeServerHandler<S, R>
 where
     S: Store + 'static,
+    R: ResponseSender,
 {
     pub(super) fn new(store: Arc<S>) -> Self {
-        ExchangeServerHandler { _store: store }
+        let (tx, rx) = mpsc::channel(32);
+        ExchangeServerHandler { store, rx, tx }
     }
 
-    #[instrument(level = "trace", skip(self, _respond_to))]
-    pub(super) fn on_request_received(
+    #[instrument(level = "trace", skip(self, response_channel))]
+    pub(super) fn on_request_received<Id>(
         &mut self,
         peer: PeerId,
-        request_id: RequestId,
+        request_id: Id,
         request: HeaderRequest,
-        _respond_to: ResponseChannel<Vec<HeaderResponse>>,
-    ) {
-        // TODO
+        response_channel: R::Channel,
+    ) where
+        Id: Display + Debug,
+    {
+        let Some((amount, data)) = parse_request(request) else {
+            self.handle_invalid_request(response_channel);
+            return;
+        };
+
+        match data {
+            header_request::Data::Origin(0) => {
+                self.handle_request_current_head(response_channel);
+            }
+            header_request::Data::Origin(height) => {
+                self.handle_request_by_height(response_channel, height, amount);
+            }
+            header_request::Data::Hash(hash) => {
+                self.handle_request_by_hash(response_channel, hash);
+            }
+        };
     }
 
-    #[instrument(level = "trace", skip(self))]
     pub(super) fn on_response_sent(&mut self, peer: PeerId, request_id: RequestId) {
-        // TODO
+        trace!("response_sent; request_id: {request_id}, peer: {peer}");
     }
 
-    #[instrument(level = "trace", skip(self))]
     pub(super) fn on_failure(
         &mut self,
         peer: PeerId,
         request_id: RequestId,
         error: InboundFailure,
     ) {
-        // TODO
+        // TODO: cancel job if libp2p already failed it?
+        trace!("on_failure; request_id: {request_id}, peer: {peer}, error: {error:?}");
+    }
+
+    pub fn poll(&mut self, cx: &mut Context<'_>, sender: &mut R) -> Poll<()> {
+        if let Poll::Ready(Some((channel, response))) = self.rx.poll_recv(cx) {
+            sender.send_response(channel, response);
+            return Poll::Ready(());
+        }
+
+        Poll::Pending
+    }
```

Comment:
Since this method doesn't return an actual result, the loop pattern needs to be used

Reply:
Fixed properly according to libp2p guidelines :+1: